### PR TITLE
stats: add library filter

### PR DESCRIPTION
**Pull request by jma**
* Adds filter by library properties in the stats configuration.
* Adds elasticsearch filter by library filters.
* Creates an indicator configuration.

Co-Authored-by: Johnny Mariéthoz <Johnny.Mariethoz@rero.ch>

### CODE REVIEW

File: rero_ils/modules/stats/api/indicators/circulation.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+from invenio_search.api import RecordsSearch
+
+from rero_ils.modules.loans.logs.api import LoanOperationLog
+from rero_ils.modules.locations.api import LocationsSearch
+
+from .base import IndicatorCfg
+
+
+class NumberOfCirculationCfg(IndicatorCfg):
+    """Number of deleted items."""
```
Comment (PascalRepond) [inline rero_ils/modules/stats/api/indicators/circulation.py:32]: Incorrect docstring

File: rero_ils/modules/stats/api/indicators/others.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                f'{self.cfg.locations[bucket.key]} ({bucket.key})',
+            'type': lambda: bucket.key,
+            'created_month': lambda: bucket.key_as_string,
+            'created_year': lambda: bucket.key_as_string
+        }
+        return cfg[distribution]()
+
+
+class NumberOfDeletedItemsCfg(IndicatorCfg):
+    """Number of deleted items."""
```
Comment (PascalRepond) [inline rero_ils/modules/stats/api/indicators/others.py:242]: Incorrect docstring

File: rero_ils/modules/stats/api/indicators/others.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+            'library': lambda:
+                f'{self.cfg.libraries[bucket.key]} ({bucket.key})',
+            'action_month': lambda: bucket.key_as_string,
+            'action_year': lambda: bucket.key_as_string
+        }
+        return cfg[distribution]()
+
+
+class NumberOfILLRequests(IndicatorCfg):
+    """Number of deleted items."""
```
Comment (PascalRepond) [inline rero_ils/modules/stats/api/indicators/others.py:307]: Same

**Review by PascalRepond — APPROVED** [review]

File: rero_ils/modules/stats/api/indicators/__init__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019-2022 RERO
```
Comment (lauren-d) [inline rero_ils/modules/stats/api/indicators/__init__.py:4]: 2019-2023 ?

**Review by lauren-d — APPROVED** [review]

File: rero_ils/modules/stats/api/indicators/circulation.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        self.trigger = trigger
+        super().__init__(report_cfg)
+
+    @property
+    def query(self):
+        """Base Elasticsearch Query.
+
+        :returns: an elasticsearch query object
+        """
+        es_query = RecordsSearch(index=LoanOperationLog.index_name)[:0]\
```
Comment (rerowep) [inline rero_ils/modules/stats/api/indicators/circulation.py:49]: Why not using `LoanOperationLogsSearch`?

**Comment by PascalRepond** [comment]
## Tests (round 1)

- [ ] Missing owning library information in config (I can edit and delete any config in the organization)
    - [ ] By default the config list should be filtered by my current library
- [ ] Config list: there should be a library and a frequency facet
- [ ] Config list: The config brief view should display the indicator (and frequency)
- [ ] not blocking: I cannot delete a config if a report exists, it says "cannot delete. undefined reports". Deleting a config should delete all linked reports (but it maybe needs a warning)
- [ ] I cannot see a legacy librarian (even with stat permissions) stat "authorization required"

File: rero_ils/modules/stats/api/indicators/others.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+class NumberOfDeletedItemsCfg(IndicatorCfg):
+    """Number of deleted items."""
+
+    @property
+    def query(self):
+        """Base Elasticsearch Query.
+
+        :returns: an elasticsearch query object
+        """
+        es_query = RecordsSearch(index=LoanOperationLog.index_name)[:0]\
```
Comment (rerowep) [inline rero_ils/modules/stats/api/indicators/others.py:250]: See above.

**Review by rerowep — APPROVED** [review]

File: rero_ils/modules/stats/api/indicators/base.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    @abstractmethod
+    def query(self):
+        """Base Elasticsearch Query.
+
+        :returns: an elasticsearch query object
+        """
+        raise NotImplementedError
+
+    @property
+    def aggregation(self, distribution):
```
Comment (zannkukai) [inline rero_ils/modules/stats/api/indicators/base.py:44]: Add `@abstractmethod` to be consistent with `def query()`.

File: rero_ils/modules/stats/api/indicators/base.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        """Elasticsearch Aggregation configuration to compute distributions.
+
+        :param distrubtion: str - report distrubtion name
+        :returns: an elasticsearch aggregation object
+        """
+        raise NotImplementedError
+
+    def label(self, distribution, bucket):
```
Comment (zannkukai) [inline rero_ils/modules/stats/api/indicators/base.py:52]: `@abstractmethod` too

**Review by zannkukai — APPROVED** [review]

**Comment by jma** [comment]
> * [x]  Config list: The config brief view should display the indicator (and frequency)

I would not add the frequency as it can be filtered using facets.

> * [ ]  not blocking: I cannot delete a config if a report exists, it says "cannot delete. undefined reports". Deleting a config should delete all linked reports (but it maybe needs a warning)

A new active filter has been added to the facet to show only the active stats. This can replace this functionality.

**Pull request merged**